src/game/Server/WorldSocket.cpp

```diff
@@ -150,143 +159,102 @@ bool WorldSocket::Open()
     {
-        m_useExistingHeader = false;
-        header = m_existingHeader;
-
-        ReadSkip(sizeof(ClientPktHeader));
-    }
-    else
-    {
-        if (!Read((char*)&header, sizeof(ClientPktHeader)))
-        {
-            errno = EBADMSG;
-            return false;
-        }
-
+        if (error) return;
         // thread safe due to always being called from service context
-        m_crypt.DecryptRecv((uint8*)&header, sizeof(ClientPktHeader));
-
-        EndianConvertReverse(header.size);
-        EndianConvert(header.cmd);
-    }
-
-    // there must always be at least four bytes for the opcode,
-    // and 0x2800 is the largest supported buffer in the client
-    if ((header.size < 4) || (header.size > 0x2800) || (header.cmd >= NUM_MSG_TYPES))
-    {
-        sLog.outError("WorldSocket::ProcessIncomingData: client sent malformed packet size = %u , cmd = %u", header.size, header.cmd);
-
-        errno = EINVAL;
-        return false;
-    }
-
-    // the minus four is because we've already read the four byte opcode value
-    const uint16 validBytesRemaining = header.size - 4;
-
-    // check if the client has told us that there is more data than there is
-    if (validBytesRemaining > ReadLengthRemaining())
-    {
-        // we must preserve the decrypted header so as not to corrupt the crypto state, and to prevent duplicating work
-        m_useExistingHeader = true;
-        m_existingHeader = header;
-
-        // we move the read pointer backward because it will be skipped again later.  this is a slight kludge, but to solve
-        // it more elegantly would require introducing protocol awareness into the socket library, which we want to avoid
-        ReadSkip(-static_cast<int>(sizeof(ClientPktHeader)));
-
-        errno = EBADMSG;
-        return false;
-    }
-
-    const Opcodes opcode = static_cast<Opcodes>(header.cmd);
+        self->m_crypt.DecryptRecv((uint8*)header.get(), sizeof(ClientPktHeader));
 
-    if (IsClosed())
-        return false;
+        EndianConvertReverse(header->size);
+        EndianConvert(header->cmd);
 
-    std::unique_ptr<WorldPacket> pct(new WorldPacket(opcode, validBytesRemaining));
+        const Opcodes opcode = static_cast<Opcodes>(header->cmd);
 
-    if (validBytesRemaining)
-    {
-        pct->append(InPeak(), validBytesRemaining);
-        ReadSkip(validBytesRemaining);
-    }
+        size_t packetSize = header->size - 4;
+        std::shared_ptr<std::vector<uint8>> packetBuffer = std::make_shared<std::vector<uint8>>(packetSize);
 
-    if (sPacketLog->CanLogPacket() && IsLoggingPackets())
-        sPacketLog->LogPacket(*pct, CLIENT_TO_SERVER, GetRemoteIpAddress(), GetRemotePort());
+        self->Read(reinterpret_cast<char*>(packetBuffer->data()), packetBuffer->size(), [self, packetBuffer, opcode, packetSize](const boost::system::error_code& error, std::size_t read) -> void
+        {
+            std::unique_ptr<WorldPacket> pct(new WorldPacket(opcode, packetSize));
```

Comment:
Use ```make_unique```.

Reply:
I didn't mean to say it was not valid - it certainly is. The main benefit is indeed a more concise construction by not having to repeat the type twice, it's not another lonely 'new' in the code (always somewhat suspicious hehe). In the case of ```std::shared_ptr``` a ```make_shared``` brings more benefits (see link I also put in another comment), so I find it easier and more consistent to always type ```make_unique```/```make_shared``` without having to think much.
There is also exception safety in more complex cases, but that doesn't apply here. For some details here, image you have some function ```void someCall(std::unique_ptr<Type1>, std::unique_ptr<Type2>);```, but you call it like ```someCall(std::unique_ptr<Type1>(new Type1(...)), std::unique_ptr<Type2>(new Type2(...)));```. Oops, because there is no guarantee that the unique_ptr objects will be made before any exception is thrown (say by the second allocation or even a construction after both allocations happened), so you may actually leak doing so. Using ```make_unique``` there's no such problem. So what's not to like, right?

Reply:
I will use std::unique_ptr<WorldPacket> data = std::make_unique<WorldPacket>(); because for reading its better. But good remark either way.